src/server/status_server/mod.rs

```diff
@@ -373,68 +311,64 @@ where
             Some(val) => match val.parse() {
                 Ok(val) => val,
                 Err(err) => {
-                    return Box::new(ok(StatusServer::err_response(
+                    return Ok(StatusServer::err_response(
                         StatusCode::BAD_REQUEST,
                         err.to_string(),
-                    )));
+                    ));
                 }
             },
             None => 99, // Default frequency of sampling. 99Hz to avoid coincide with special periods
         };
 
         let prototype_content_type: hyper::http::HeaderValue =
             hyper::http::HeaderValue::from_str("application/protobuf").unwrap();
-        Box::new(
-            Self::dump_rsprof(seconds, frequency)
-                .and_then(move |report| {
-                    let mut body: Vec<u8> = Vec::new();
-                    if req.headers().get("Content-Type") == Some(&prototype_content_type) {
-                        match report.pprof() {
-                            Ok(profile) => match profile.encode(&mut body) {
-                                Ok(()) => {
-                                    info!("write report successfully");
-                                    Box::new(ok(StatusServer::err_response(StatusCode::OK, body)))
-                                }
-                                Err(err) => Box::new(ok(StatusServer::err_response(
-                                    StatusCode::INTERNAL_SERVER_ERROR,
-                                    err.to_string(),
-                                ))),
-                            },
-                            Err(err) => Box::new(ok(StatusServer::err_response(
-                                StatusCode::INTERNAL_SERVER_ERROR,
-                                err.to_string(),
-                            ))),
-                        }
-                    } else {
-                        match report.flamegraph(&mut body) {
-                            Ok(_) => {
-                                info!("write report successfully");
-                                Box::new(ok(StatusServer::err_response(StatusCode::OK, body)))
-                            }
-                            Err(err) => Box::new(ok(StatusServer::err_response(
-                                StatusCode::INTERNAL_SERVER_ERROR,
-                                err.to_string(),
-                            ))),
-                        }
+        let report = match Self::dump_rsprof(seconds, frequency).await {
+            Ok(report) => report,
+            Err(err) => {
+                return Ok(StatusServer::err_response(
```

Comment:
Can this pattern be reduced?

Reply:
I cannot find a good idea. Both `Option` and `Result` use this pattern. I failed to find a unified way to handle them.
If the return type is `std::result::Result<Response, Response>`, we could use `?` for early return but some method needs `hyper::Result`. I want to make them look consistent.